Adding the option for specify the project title on packaged project

SeqLaz opened the pull request:
This adds the option to specify a new project name and title in the exported project when packaging.

| Before | After |
|--------|-------|
| ![Before Image](https://github.com/user-attachments/assets/543f0fc9-4034-463e-a638-1a0b453281fc) | ![After Image](https://github.com/user-attachments/assets/9ad7c145-a6b8-454a-98a3-496c6ae21ef9) |

Related PR: https://github.com/opengisch/libqfieldsync/pull/94

suricactus commented on qfieldsync/ui/package_dialog.ui:
```diff
@@ -22,15 +22,15 @@
       <widget class="QLabel" name="label_2">
        <property name="text">
-        <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-weight:600;&quot;&gt;Project:&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+        <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-weight:600;&quot;&gt;Exported Project Title:&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
This should not be HTML.

SeqLaz replied:
Good catch!

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -189,6 +192,8 @@ def package_project(self):
             QApplication.setOverrideCursor(Qt.WaitCursor)
             offline_convertor.convert()
+            self.project.setTitle(self.original_title_project)
+            open_project(self.project.fileName())
```
I guess you don't need it in this PR, do you?

SeqLaz replied:
Due this was removed on the other [PR](https://github.com/opengisch/qfieldsync/pull/616) now this will be necessary there

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -73,7 +73,8 @@ def __init__(self, iface, project, offline_editing, parent=None):
         self.dirsToCopyWidget = DirsToCopyWidget()
         self.__project_configuration = ProjectConfiguration(self.project)
-        self.project_lbl.setText(get_project_title(self.project))
+        self.original_title_project = project.title()
```
```suggestion
        self.original_project_title = project.title()
```

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -73,7 +73,8 @@ def __init__(self, iface, project, offline_editing, parent=None):
         self.__project_configuration = ProjectConfiguration(self.project)
-        self.project_lbl.setText(get_project_title(self.project))
+        self.original_title_project = project.title()
+        self.new_project_title.setText(self.original_title_project)
```
```suggestion
        self.packaged_project_title.setText(self.original_title_project)
```

suricactus commented on qfieldsync/ui/package_dialog.ui:
```diff
@@ -7,42 +7,61 @@
+     <property name="text">
+      <string>&lt;html&gt;&lt;head/&gt;&lt;body&gt;&lt;p&gt;&lt;span style=&quot; font-weight:600;&quot;&gt;Packaged Project Name&lt;/span&gt;&lt;/p&gt;&lt;/body&gt;&lt;/html&gt;</string>
+     </property>
+     <property name="textFormat">
+      <enum>Qt::RichText</enum>
+     </property>
```
```suggestion
```

I guess the HTML problem above comes from the fact that the QLabel is `RichText`.

suricactus commented on qfieldsync/ui/package_dialog.ui:
```diff
@@ -7,42 +7,61 @@
+      <string>&lt;p&gt;&lt;span style=&quot; font-weight:600;&quot;&gt;Packaged Project Title&lt;/span&gt;&lt;/p&gt;</string>
+     </property>
+     <property name="textFormat">
+      <enum>Qt::RichText</enum>
```
Same as above.

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -70,10 +70,15 @@ def __init__(self, iface, project, offline_editing, parent=None):
         self.dirsToCopyWidget = DirsToCopyWidget()
         self.__project_configuration = ProjectConfiguration(self.project)
-        self.project_lbl.setText(get_project_title(self.project))
+        self.original_project_title = project.title()
+        self.packaged_project_title.setText(self.original_project_title)
```
```suggestion
        self.packaged_project_title.setText(project.title())
```

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -70,10 +70,15 @@ def __init__(self, iface, project, offline_editing, parent=None):
         self.iface = iface
         self.offliner = QgisCoreOffliner(offline_editing=offline_editing)
         self.project = project
+        self.original_project_path = self.project.fileName()
```
This is only for internal use, right? Can we make it a "private" property by prefixing with underscore?
```suggestion
        self._original_project_path = self.project.fileName()
```

Also, please move it next to the rest of the changes in this PR a few lines below.

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -70,10 +70,15 @@ def __init__(self, iface, project, offline_editing, parent=None):
-        self.project_lbl.setText(get_project_title(self.project))
+        self.original_project_title = project.title()
+        self.packaged_project_title.setText(self.original_project_title)
+        self.packaged_project_name.setText(self.project.baseName())
```
What if the project we are trying to package is not saved yet? Imagine you open a new blank project, set all the layers, config etc and then press "package project" without ever saving it to the disk?

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -70,10 +70,15 @@ def __init__(self, iface, project, offline_editing, parent=None):
+        self.original_project_title = project.title()
+        self.packaged_project_title.setText(self.original_project_title)
+        self.packaged_project_name.setText(self.project.baseName())
+        self.tmp_project = make_temp_qgis_file(self.project)
```
```suggestion
        self.tmp_project_filename = make_temp_qgis_file(self.project)
```

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -168,6 +173,14 @@ def package_project(self):
         self.qfield_preferences.set_value("exportDirectoryProject", export_folder)
         self.dirsToCopyWidget.save_settings()
 
+        tmp_path = os.path.dirname(self.tmp_project)
```
```suggestion
        tmp_project_filepath = Path(self.tmp_project_filename).parent
```

Or whatever is the syntax using `pathlib`. Note that in all the codebase we prefer using `pathlib` over `os.path`.

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -189,6 +202,7 @@ def package_project(self):
         try:
             QApplication.setOverrideCursor(Qt.WaitCursor)
             offline_convertor.convert()
+            open_project(self.original_project_path)
```
I think this is already handled internally in the `OfflineConverter`. It should reopen the original project internally, no? If yes, then we need to make sure the OfflineConverter opens the right project.

SeqLaz replied:
This is done in this [PR](https://github.com/opengisch/libqfieldsync/pull/94)

suricactus reviewed (commented):
Instead of complicating the code in the caller (in this case QFieldSync), we should add this functionality within the OfflineConverter itself. Why? Because the converter already does the copy of the file.

For example here we have hardcoded the name of the output file: https://github.com/opengisch/libqfieldsync/blob/master/libqfieldsync/offline_converter.py#L276

What we can do is to pass the output name to be a parameter when we construct the `OfflineConverter` class. We can even replace the `export_dir` param with `export_filename` param, and then derive the `export_dir` to be the path to `export_filename`.

suricactus commented:
Please note that most of the suggestion here need code-wide changes, the suggestions are only pointing to the highlights.

SeqLaz commented:
@suricactus The changes and suggestions were addressed in this [PR](https://github.com/opengisch/libqfieldsync/pull/94)

SeqLaz commented:
> Please note that most of the suggestion here need code-wide changes, the suggestions are only pointing to the highlights.

@suricactus The changes and suggestions were addressed in this [PR](https://github.com/opengisch/libqfieldsync/pull/94)

suricactus commented:
Few comments after the libqfieldsync review:
- Drop the "export directory" and export finame in favor of single "export filename" with the full path
- After clicking "Create" OR when changing the value of "export filename", make sure the user is notified there is already a file with that name in that location.
- Prepopulate the "export_filename" with the default value of "export directory" with "/" and current project basename and "_qfield" suffix and ".qgs" extension. You can create a utility function in libqfieldsync `get_export_filename_suggestion(project.fileName())`

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -145,15 +149,43 @@ def get_export_folder_from_dialog(self):
         self.button_box.button(QDialogButtonBox.Save).setEnabled(False)
 
-        export_folder = self.get_export_folder_from_dialog()
+        export_packaged_project = Path(self.packaged_project_filename.text())
```
```suggestion
        packaged_project_filename = Path(self.packaged_project_filename.text())
```

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -145,15 +149,43 @@ def get_export_folder_from_dialog(self):
+        if export_packaged_project.exists():
+            reply = QMessageBox.question(
+                self,
+                "File Already Exists",
+                f"'{export_packaged_project.name}' already exists. Do you want to overwrite it?",
```
```suggestion
                self.tr("File Already Exists"),
                self.tr('File "{}" already exists. Do you want to overwrite it?').format(export_packaged_project.name),
```
don't forget the translation

suricactus commented on qfieldsync/ui/package_dialog.ui:
```diff
@@ -111,11 +81,33 @@
          <layout class="QGridLayout" name="gridLayout">
           <item row="0" column="0">
-           <widget class="QLineEdit" name="manualDir"/>
+           <widget class="QLineEdit" name="packaged_project_title"/>
```
```suggestion
           <widget class="QLineEdit" name="packagedProjectTitleLineEdit"/>
```
Note that the names here are camelCase for historical reasons. Don't mix multiple naming conventions.

suricactus commented on qfieldsync/ui/package_dialog.ui:
```diff
@@ -111,11 +81,33 @@
+         </widget>
+       </item>
+          <item row="0" column="0">
+           <widget class="QLineEdit" name="packaged_project_filename"/>
```
```suggestion
           <widget class="QgsFileWidget" name="packagedProjectFileWidget"/>
```

suricactus commented on qfieldsync/ui/package_dialog.ui:
```diff
@@ -111,11 +81,33 @@
+          <string>Packaged Project Filename</string>
+         </property>
+         <layout class="QGridLayout" name="gridLayout">
+       <item>
+         <widget class="QLineEdit" name="manualDir">
+           <property name="maximumSize">
+             <size>
+             <width>0</width>
+             <height>0</height>
+             </size>
+           </property>
+         </widget>
+       </item>
```
Why do we have the `manualDir` here? Also I think there is some kind of an error with the XML, or I miss somehow the closing `</item>` for line 93. How did you edit this?

suricactus reviewed (changes requested):
Not there yet.

SeqLaz commented:
> Not there yet.

Hey @suricactus please have a look.
I squashed due to the PR had a lot of merge. In the last commit I addressed all the suggestion and some improvements.

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -140,20 +136,51 @@ def setup_gui(self):
+            )
+        return QDir.toNativeSeparators(str(export_dirname))
+
+    def get_export_filename_suggestion(self):
```
```suggestion
    def get_export_filename_suggestion(self) -> str:
```

suricactus replied:
```suggestion
    def get_export_filename_suggestion(self) -> str:
```

Please add typing on all new methods you add.

SeqLaz commented:
Hey @suricactus I added the typing hints on the new functions I added, have a look!

suricactus commented on qfieldsync/gui/package_dialog.py:
```diff
@@ -140,20 +136,51 @@ def setup_gui(self):
+        )
+        return str(full_project_name_suggestion)
+
+    def set_export_filename_suggested(self) -> None:
+        self.packagedProjectFileWidget.setFilePath(
+            self.get_export_filename_suggestion()
+        )
```
I failed to find who uses this? In any case, such functions should not exist, it's better to do it explicitly what you do in the function body.

SeqLaz replied:
You are completely right, the function is not needed!

SeqLaz commented:
Hey @suricactus, I deleted the unnecessary function and joined the functions of export_folder and get_filename_suggestion.

suricactus reviewed (approved):
Great work!

suricactus reviewed (changes requested):
Ah, just realized, you never bumped libqfieldsync here. Last step before merging. I guess the tests will fail here too.

suricactus commented:
@SeqLaz you have a conflict now. I will resolve it to unblock the tests.

SeqLaz commented:
Hey @suricactus I've fixed the tests!.
